File: HMCLCore/src/main/java/org/jackhuang/hmcl/util/IOUtils.java
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2018  huangyuhui <huanghongxun2008@126.com>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
package org.jackhuang.hmcl.util;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.Charset;

/**
 *
 * @author huang
 */
public final class IOUtils {

    private IOUtils() {
    }

    public static final int DEFAULT_BUFFER_SIZE = 8 * 1024;

    public static byte[] readFullyWithoutClosing(InputStream stream) throws IOException {
        ByteArrayOutputStream result = new ByteArrayOutputStream();
        copyTo(stream, result);
        return result.toByteArray();
    }

    public static ByteArrayOutputStream readFully(InputStream stream) throws IOException {
        try (InputStream is = stream) {
            ByteArrayOutputStream result = new ByteArrayOutputStream();
            copyTo(is, result);
            return result;
        }
    }

    public static byte[] readFullyAsByteArray(InputStream stream) throws IOException {
        return readFully(stream).toByteArray();
    }

    public static String readFullyAsString(InputStream stream) throws IOException {
        return readFully(stream).toString();
    }

    public static String readFullyAsString(InputStream stream, Charset charset) throws IOException {
        return readFully(stream).toString(charset.name());
    }

    public static void copyTo(InputStream src, OutputStream dest) throws IOException {
        copyTo(src, dest, new byte[DEFAULT_BUFFER_SIZE]);
    }

    public static void copyTo(InputStream src, OutputStream dest, byte[] buf) throws IOException {
        while (true) {
            int len = src.read(buf);
            if (len == -1)
                break;
            dest.write(buf, 0, len);
        }
    }
}

File: HMCLCore/src/main/kotlin/org/jackhuang/hmcl/util/IOUtils.kt
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2017  huangyuhui <huanghongxun2008@126.com>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
package org.jackhuang.hmcl.util

import java.io.*
import java.nio.charset.Charset

const val MAX_BUFFER_SIZE = 4096

fun Closeable.closeQuietly() {
    try {
        this.close()
    } catch (ex: IOException) {}
}

fun InputStream.readFully(): ByteArrayOutputStream {
    try {
        val ans = ByteArrayOutputStream()
        copyTo(ans)
        return ans
    } finally {
        this.closeQuietly()
    }
}

fun InputStream.readFullyAsByteArray(): ByteArray =
        readFully().toByteArray()

fun InputStream.readFullyAsString(): String =
        readFully().toString()

fun InputStream.readFullyAsString(charset: Charset): String =
        readFully().toString(charset.name())

fun InputStream.copyTo(dest: OutputStream, buf: ByteArray) {
    while (true) {
        val len = read(buf)
        if (len == -1)
            break
        dest.write(buf, 0, len)
    }
}

fun InputStream.copyToAndClose(dest: OutputStream) {
    this.use { input ->
        dest.use { output ->
            input.copyTo(output)
        }
    }
}

/**
 * @param cmd the command line
 * @return the final command line
 */
fun makeCommand(cmd: List<String>): String {
    val cmdbuf = StringBuilder(120)
    for (i in cmd.indices) {
        if (i > 0)
            cmdbuf.append(' ')
        val s = cmd[i]
        if (s.indexOf(' ') >= 0 || s.indexOf('\t') >= 0)
            if (s[0] != '"') {
                cmdbuf.append('"')
                cmdbuf.append(s)
                if (s.endsWith("\\"))
                    cmdbuf.append("\\")
                cmdbuf.append('"')
            } else if (s.endsWith("\""))
                // The argument has already been quoted.
                cmdbuf.append(s)
            else
                // Unmatched quote for the argument.
                throw IllegalArgumentException()
        else
            cmdbuf.append(s)
    }

    return cmdbuf.toString()
}